### PR TITLE
Adds a reagent injector component and BCI manipulators to all circuit labs

### PR DESCRIPTION
**Pull request by RikuTheKiller**
## About The Pull Request

This PR adds a reagent injector component that's exclusive to BCIs. (Requested to be integrated into BCIs by Mothblocks.)
When outside of a circuit, the component itself stores the reagents. However, if it's inside of a BCI, the storage is moved to the BCI. The storage can contain up to 15u of reagents and acts like an open container. (However, it won't spill even if you throw it, it just acts like an open container code-wise, don't worry about it.)
You can only have one reagent injector in a circuit. Trying to insert multiple will give you an error message.
The entire dose is administered at once. (Requirement set by Mothblocks.)

Please don't try to dispute any of the specific limitations in the comments as they're out of my control. They're reasonable anyways.

Reagent Injector Input/Output:
Inject (Input Signal) - Administers all reagents currently stored inside of the BCI into the user.
Injected (Output Signal) - Triggered when reagents are injected. Not triggered if the reagent storage is empty.

New BCI Input:
Show Charge Meter (Number) - Toggles showing the charge meter action. (Adds some capacity for stealth.)

Install Detector Outputs: (Added following a comment about having to use weird workarounds for proper loops.)
Current State (Number) - Outputs 1 if the BCI is implanted and 0 if it's not.
Installed (Signal) - Triggered when the BCI is implanted into it's user.
Removed (Signal) - Triggered when the BCI is removed from it's user.

This PR also adds BCI manipulation chambers to all currently present circuit labs. (Solution proposed by Mothblocks.)
Yes I had to do some other mapping changes to allow for this. No I don't have any mapping experience, why do you ask?

## Why It's Good For The Game

One small step for BCIs, one giant leap for circuit kind. (First "proper" circuit to human interaction in the entire game!)

This allows for some funky stuff and also makes it less of a pain in the ass to use BCIs. What's not to love?

## Changelog

:cl:
add: Added a reagent injector component and BCI manipulators to all circuit labs. (+ install detector component)
/:cl:

### CODE REVIEW

File: code/modules/wiremod/shell/brain_computer_interface.dm [hunk 1]
```diff
@@ -10,6 +10,8 @@
 /obj/item/organ/internal/cyberimp/bci/Initialize(mapload)
 	. = ..()
 
+	create_reagents(15, TRANSPARENT | REFILLABLE) // This is used for the reagent injector component.
```
Comment (Mothblocks) [inline code/modules/wiremod/shell/brain_computer_interface.dm:13]: The reagent_injector component itself should add this

Comment (Mothblocks) [reply code/modules/wiremod/shell/brain_computer_interface.dm:13]: Then you could even transfer the reagents between different BCIs because when you take out the component it could move the reagents with it

File: code/modules/research/techweb/all_nodes.dm [hunk 2]
```diff
@@ -826,6 +826,7 @@
 		"comp_bar_overlay",
 		"comp_bci_action",
 		"comp_counter_overlay",
+		"comp_reagent_injector",
```
Comment (Mothblocks) [inline code/modules/research/techweb/all_nodes.dm:829]: ABC order--r comes after o.

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 1 above)
Comment (RikuTheKiller) [reply code/modules/wiremod/shell/brain_computer_interface.dm:13]: its intentionally made this way so that you can add multiple reagent injector components without any conflicts

File: code/modules/research/techweb/all_nodes.dm (hunk 2 above)
Comment (RikuTheKiller) [reply code/modules/research/techweb/all_nodes.dm:829]: god mcfucking damnit

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 1 above)
Comment (Mothblocks) [reply code/modules/wiremod/shell/brain_computer_interface.dm:13]: Multiple reagent injector components should be using the same reagents, and it should only exist if the reagent container is actually added. This doesn't conflict with what I want.

**Comment by private-tristan** [comment]
tfw you're critted by a nukie but your BCI beeps  "Major fracture detected, Automatic medical system engaged, Morphine administered"

**Comment by Mothblocks** [comment]
If it fits my requirements yes

**Comment by moocowswag** [comment]
Yay maint circuit chem base buff.

**Comment by RikuTheKiller** [comment]
> tfw you're critted by a nukie but your BCI beeps "Major fracture detected, Automatic medical system engaged, Morphine administered"
> 
> also how do you refill it? do you need to take it out, syringe/beaker it, and then stick it back in you?

yes, you do need to take it out (cough cough Mothblocks)

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 1 above)
Comment (RikuTheKiller) [reply code/modules/wiremod/shell/brain_computer_interface.dm:13]: > Multiple reagent injector components should be using the same reagents, and it should only exist if the reagent container is actually added. This doesn't conflict with what I want.

so like, wouldnt this still make you take out the component and whatnot? i just dont feel like making you reroute your circuitry just to refill it, ya know?

unless you feel like making me overload all the reagent procs to the BCI... in which case, why bother?

Comment (Mothblocks) [reply code/modules/wiremod/shell/brain_computer_interface.dm:13]: No it would not because it would be creating the reagents on the BCI

**Comment by moocowswag** [comment]
> tfw you're critted by a nukie but your BCI beeps "Major fracture detected, Automatic medical system engaged, Morphine administered"
> 
Believe it or not, this isnt something you could easily automate for long term, health detection needs to be pulsed and then your health needs to be compared to whatever threshold you want it to activate at.

a clock component would run out of power in the span of minutes which is why youd need either a button that turns the clock on/off or just a button that injects yourself with the chems...

a slow clock using a delay component would consume less power but that needs to be pulsed once (and only once) after install to work which is why its iffy, if there was bci component that outputed a pulse on install/uninstall this would be a lot more user friendly and it wouldnt be limited to just me doing work arounds for it while the vast majority of players that even try to make this will probably run out of power and be sad that all the hoops they had to jump through to get a combination of chems and circuits was for nothing.

TLDR this puts another massive skill ceiling on circuits while the floor is nonexistent due to weird work arounds that most players wouldnt bother doing.

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 1 above)
Comment (RikuTheKiller) [reply code/modules/wiremod/shell/brain_computer_interface.dm:13]: > 

All atoms have reagents datums, what's the point here? Most reagents datums are created at initialize, too. I also need some not-shit way of detecting if another component has already added it. Removing it isn't feasible, don't want to be deleting people's reagents.

With all this said, the current method is the most sane way of doing it.

Why does it even matter anyways? Do you care that much about a "It Contains:" message or why are you so adamant about it?

**Comment by RikuTheKiller** [comment]
> 

It is meant to be situational, but I'll draft the PR till I make a BCI installation detector. Thanks for the idea.

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 1 above)
Comment (Mothblocks) [reply code/modules/wiremod/shell/brain_computer_interface.dm:13]: A circuit that doesn't have a reagents component inside it should not have its own reagents container, I don't know why that's such a big deal

Comment (RikuTheKiller) [reply code/modules/wiremod/shell/brain_computer_interface.dm:13]: > A circuit that doesn't have a reagents component inside it should not have its own reagents container, I don't know why that's such a big deal. It's silly UX otherwise, it would mean yes the description gives a useless line of text, but also that you would be able to inject reagents into something that it won't do anything with.

Give me a way of handling it without shitcode please, don't want removing the component to delete any reagents.

Comment (Mothblocks) [reply code/modules/wiremod/shell/brain_computer_interface.dm]: Give the reagents component to the circuit component

When added to a BCI, transfer the reagents component if it does not already have one

If the component is taken out, and it is the last one that does not have reagents in it, transfer the reagents component to that

**Comment by optimumtact** [comment]
it's a bad idea mothblocks

**Comment by Mothblocks** [comment]
oh yeah that reminds me this needs logs

File: code/modules/wiremod/components/bci/reagent_injector.dm [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+	inject = add_input_port("Inject", PORT_TYPE_SIGNAL, trigger = PROC_REF(trigger_inject))
+	injected = add_output_port("Injected", PORT_TYPE_SIGNAL)
+
+/obj/item/circuit_component/reagent_injector/proc/trigger_inject()
+	CIRCUIT_TRIGGER
+	if(!bci.owner)
+		return
+	if(bci.owner.reagents.total_volume + bci.reagents.total_volume > bci.owner.reagents.maximum_volume)
+		return
+	var/units = bci.reagents.trans_to(bci.owner.reagents, bci.reagents.total_volume, methods = INJECT)
```
Comment (Mothblocks) [inline code/modules/wiremod/components/bci/reagent_injector.dm:36]: Add a `log_attack` (I think?? CC @dragomagol on what fits here best) to log that an injection happened, with who, and what reagents.

**Comment by moocowswag** [comment]
if you want to you could lock it as an implant (must be researched with other implants) that you can attach to the bci?

or not even attach it at all and just make it triggerable with a signaler.

File: code/modules/wiremod/components/bci/reagent_injector.dm (hunk 3 above)
Comment (dragomagol) [reply code/modules/wiremod/components/bci/reagent_injector.dm:36]: Yup, `log_combat` is the typical reagent transfer logging proc

Comment (RikuTheKiller) [reply code/modules/wiremod/components/bci/reagent_injector.dm:36]: Sure, will do once I'm not busy.

**Comment by RikuTheKiller** [comment]
> if you want to you could lock it as an implant (must be researched with other implants) that you can attach to the bci?
> 
> or not even attach it at all and just make it triggerable with a signaler.
> 
> Anyways even I am probably gonna struggle making this work well (in terms of powergaming) despite the limitations but I think I can make something that is "gimmicky" and gives a "wow" factor that might get people to use it just cause its cool.
> 
> 15u isnt enough for any of the...stam/stuncombat tricks and I cant see an obvious way to do anything explosive with it, you can already just have a circuit grenade in your backpack so the only real use is a limited injection into the host, n2o is getting nerfed so it shouldnt be as absurdly easy to get a chem mix that kills in 15u.
> 
> but ya if i do find something strong ill make it obvious.

I can already think of some funny things you could do with it.

**Comment by RikuTheKiller** [comment]
POV: waiting for merge

File: code/modules/wiremod/components/bci/install_detector.dm [hunk 4]
```diff
@@ -0,0 +1,50 @@
+/**
+ * # Install Detector Component
+ *
+ * Detects when a BCI is installed/removed.
+ * Requires a BCI shell.
+ */
+
+/obj/item/circuit_component/install_detector
+	display_name = "Install Detector"
+	desc = "A component that detects when a BCI is installed or removed from it's user."
```
Comment (Mothblocks) [inline code/modules/wiremod/components/bci/install_detector.dm:10]:
```suggestion
	desc = "A component that detects when a BCI is installed or removed from its user."
```

File: code/modules/wiremod/core/integrated_circuit.dm [hunk 5]
```diff
@@ -231,6 +253,11 @@ GLOBAL_LIST_EMPTY_TYPED(integrated_circuits, /obj/item/integrated_circuit)
 	if(!to_add.add_to(src))
 		return FALSE
 
+	if(to_add.circuit_flags & CIRCUIT_NO_DUPLICATES)
```
Comment (Mothblocks) [inline code/modules/wiremod/core/integrated_circuit.dm:256]: @Watermelon914 Should this be in `add_component` or `add_component_manually`? I noticed that the print function does not use the manually proc? Is that intentional?

File: code/modules/wiremod/shell/brain_computer_interface.dm [hunk 6]
```diff
@@ -101,47 +101,67 @@
 
 	var/datum/port/input/message
 	var/datum/port/input/send_message_signal
+	var/datum/port/input/show_charge_meter
 
 	var/datum/port/output/user_port
 
 	var/datum/weakref/user
 
+	var/obj/item/organ/internal/cyberimp/bci/bci
```
Comment (Mothblocks) [inline code/modules/wiremod/shell/brain_computer_interface.dm:110]: Set this to null in Destroy

File: code/modules/wiremod/shell/brain_computer_interface.dm [hunk 7]
```diff
@@ -101,47 +101,67 @@
...
+		bci.actions -= charge_action
+		QDEL_NULL(charge_action)
+
 /obj/item/circuit_component/bci_core/register_shell(atom/movable/shell)
-	var/obj/item/organ/internal/cyberimp/bci/bci = shell
+	bci = shell
 
-	charge_action = new(src)
-	bci.actions += list(charge_action)
+	show_charge_meter.set_value(TRUE) // Defaults to TRUE (1) which shows the charge meter.
```
Comment (Mothblocks) [inline code/modules/wiremod/shell/brain_computer_interface.dm:145]:
```suggestion
	show_charge_meter.set_value(TRUE)
```

This comment is not helpful, the code itself should be obvious enough

**Review by Mothblocks — CHANGES_REQUESTED** [review]

File: code/modules/wiremod/core/integrated_circuit.dm (hunk 5 above)
Comment (RikuTheKiller) [reply code/modules/wiremod/core/integrated_circuit.dm:256]: I mean, this is the proc that encompasses it all.

File: code/modules/wiremod/shell/brain_computer_interface.dm (hunk 6 above)
Comment (RikuTheKiller) [reply code/modules/wiremod/shell/brain_computer_interface.dm:110]: Why? Other components don't do this.

Comment (Mothblocks) [reply code/modules/wiremod/shell/brain_computer_interface.dm:110]: Oh right this gets removed in unregister_shell, ok its fine then
